client/app/services/parameters/Parameter.js

```diff
@@ -0,0 +1,130 @@
+import { isNull, isObject, isFunction, isUndefined, isEqual, has } from 'lodash';
+import { TextParameter, NumberParameter } from '.';
+
+class Parameter {
+  constructor(parameter, parentQueryId) {
+    this.title = parameter.title;
+    this.name = parameter.name;
+    this.type = parameter.type;
+    this.global = parameter.global; // backward compatibility in Widget service
+    this.parentQueryId = parentQueryId;
+
+    // Used for meta-parameters (i.e. dashboard-level params)
+    this.locals = [];
+
+    // Used for URL serialization
+    Object.defineProperty(this, 'urlPrefix', {
+      configurable: true,
+      enumerable: false, // don't save it
+      writable: true,
+      value: 'p_',
+    });
+  }
+
+  static create(param, parentQueryId) {
+    switch (param.type) {
+      case 'number':
+        return new NumberParameter(param, parentQueryId);
+      default:
+        return new TextParameter({ ...param, type: 'text' }, parentQueryId);
+    }
+  }
+
+  static normalizeValue(value) {
+    if (isUndefined(value)) {
+      return null;
+    }
+    return value;
+  }
+
+  static getValue(param, extra = {}) {
+    if (!isObject(param) || !isFunction(param.getValue)) {
+      return null;
+    }
+
+    return param.getValue(extra);
+  }
+
+  static setValue(param, value) {
+    if (!isObject(param) || !isFunction(param.setValue)) {
+      return null;
+    }
+
+    return param.setValue(value);
+  }
+
+  get isEmpty() {
+    return isNull(this.getValue());
+  }
+
+  get hasPendingValue() {
+    return this.pendingValue !== undefined && !isEqual(this.pendingValue, this.normalizedValue);
+  }
+
+  get normalizedValue() {
+    return this.$$value;
```

Comment:
Basically from what I understood `$$value` will be the value used internally in Inputs and `value` is the stored value in the backend (also is the one that goes to the query in Refresh schedule cases). I'm not 100% happy with how this is, but not the biggest issue here and it worked so far, perhaps just gonna see different name to the attribute.

Reply:
Totally! Can you provide a breakdown of the different value types and what format they're in?
1) param.value
2) param.$$value
3) param.getValue()
4) param.normalizedValue
5) param.pendingValue
etc.